src/query/query_executor.c

```diff
@@ -1160,6 +1182,11 @@ qexec_end_one_iteration (THREAD_ENTRY * thread_p, XASL_NODE * xasl, XASL_STATE *
 	  GOTO_EXIT_ON_ERROR;
 	}
 
+      if (XASL_IS_FLAGED (xasl, XASL_IS_MERGE_QUERY))
+	{
+	  qexec_set_upd_ins_flag (xasl);
+	}
```

Comment:
I disagree with the source code location. This seems to be a routine that should be in the XASL generator such as pt_to_merge_xasl(). If it is in end_one_iteration, it will run every row.